src/reporter/mod.rs

```diff
@@ -0,0 +1,277 @@
+use crate::commands::DEFAULT_CONFIG_PATH;
+use crate::settings::{self, toml::Manifest};
+
+use std::env::args;
+use std::fs::{read_dir, File};
+use std::io::{BufReader, Write};
+use std::panic;
+use std::panic::PanicInfo;
+use std::path::{Path, PathBuf};
+use std::{collections::HashMap, fs};
+
+use anyhow::{anyhow, Result};
+use backtrace::{Backtrace, BacktraceFrame};
+use path_slash::PathExt;
+use serde::{Deserialize, Serialize};
+use sys_info::{os_release, os_type};
+use uuid::Uuid;
+
+const PANIC_UNWIND_START_MARKER: &str = "rust_begin_unwind";
+const BACKTRACE_PATH_PREFIX: &str = "backtrace::";
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Report {
+    uuid: Uuid,
+    timestamp_ms: u128,
+    host_env: HashMap<String, String>, // "os": "..." TODO: consider struct over HashMaps
+    project_info: HashMap<String, String>,
+    args: Vec<String>,
+    panic: Option<String>,
+    location: Option<String>,
+    backtrace: String,
+}
+
+/// Overrides any panic hooks with wrangler's error reporting, which logs error reports to disl with
+/// details from a panic and useful information from the wrangler user's system for debugging.
+pub fn init() {
+    // TODO: consider using panic::take_hook, and showing the original panic to the end-user without
+    // polluting the console to the point the wrangler error report message is lost in the noise.
+    panic::set_hook(Box::new(|panic_info| generate_report(Some(panic_info))));
+}
+
+/// finds the most recent log in the error report directory, unless `log` is Some, and will try to
+/// read that file instead.
+pub fn read_log(log: Option<&str>) -> Result<Report> {
+    match log {
+        Some(path) => {
+            let r = BufReader::new(File::open(error_report_dir()?.join(path))?);
+            serde_json::from_reader(r).map_err(|e| e.into())
+        }
+        None => latest_report(),
+    }
+}
+
+/// gathers necessary error report information, and stores on disk until uploaded. the
+pub fn generate_report(panic_info: Option<&PanicInfo>) {
+    let mut report = Report {
+        uuid: Uuid::new_v4(),
+        timestamp_ms: 0,
+        host_env: load_host_info(),
+        project_info: load_project_info(),
+        args: args().collect::<Vec<_>>(),
+        panic: panic_payload(panic_info),
+        location: None,
+        backtrace: useful_frames(),
+    };
+
+    if let Some(info) = panic_info {
+        if let Some(loc) = info.location() {
+            report.location = Some(format!("{}:{}:{}", loc.file(), loc.line(), loc.column()));
```

Comment:
can you change the format from `<file>:<line>:<column>` to something easier to parse, like `<file>, Line <line> Column <column>`? If it doesn't have to be a specific format, it'd be nice to have something more intuitive.

Reply:
yea for sure.  just went with the file:line:column as the standard for how system tools specify location. but we're not integrating this output into an IDE or something, so it's probably better to be explicit for humans.

Reply:
lol it's probably just me never having seen that format before then 😅 

thank!

Reply:
no worries :) it's a sneaky little detail, but even rustc uses it in error messages, note **src/reporter/mod.rs:106:14**:

```
error: expected one of `.`, `;`, `?`, `}`, or an operator, found `::`
   --> src/reporter/mod.rs:106:14
    |
106 |         "o_0"::err_exit(format!("wrangler encountered an unrecoverable error and failed to write the report: \n{:#?}", report), 1);
    |              ^^ expected one of `.`, `;`, `?`, `}`, or an operator

error: aborting due to previous error
```

I like the more human-friendly version as you suggested for our error reporting anyway.